### DISCUSSION TITLE
Reading a velocity field from alternate software (OpenFOAM) as a variable

### DISCUSSION THREAD

**Discussion by danielwardstfc** [idaholab/moose#24784] [Q&A General] [answered]
Hi there,

I'm trying to simulate stresses and model heat transfer in MOOSE using a velocity field I have generated in OpenFOAM.  To start with, I have a velocity field in a straight pipe, and am looking to read this in to MOOSE as a variable in order to advect heat along the pipe using MOOSE.  Does anything in MOOSE already exist that would help me to do this?

Thanks,

Daniel

**Comment by GiudGiud** [comment, marked as answer]
Hello

@freiler has done this recently using this Function and UO:
https://mooseframework.inl.gov/source/functions/PiecewiseConstantFromCSV.html
https://mooseframework.inl.gov/source/userobjects/PropertyReadFile.html

You will need to use the voronoi mode and have as many voronoi regions as coordinates/points in your CSV

Guillaume

**Comment by hugary1995** [comment]
As an alternative to what @GiudGiud suggested, ANL has this wrapper app called Cardinal that couples NekRS (a CFD code) to MOOSE. @aprilnovak is the primary developer of Cardinal.

**Reply by aprilnovak** [reply]
Hi @danielwardstfc, we have done this a lot with Cardinal (https://cardinal.cels.anl.gov/), a wrapper of NekRS inside of MOOSE. If you are committed to using OpenFOAM, you can check out Cardinal's source code to see how we have wrapped NekRS inside MOOSE, and adapt to use with OpenFOAM.

Depending on your application, you could also consider using Cardinal directly. NekRS contains capabilities for RANS, LES, and DNS using the spectral element method.

Let me know if you have any questions, happy to help!

**Reply by danielwardstfc** [reply]
Hi April, thanks for your reply.  I'm not ready to move to a full coupling yet, but I'll keep this in mind if I get there.  Thanks!
